Migrate from typings to @types packages

devoto13 opened the pull request:
Allows to import this package into another TypeScript file and get proper type definitions.

blakeembrey commented:
We can't enable this until it's migrated away from Typings: https://github.com/TypeStrong/ts-node/blob/master/typings.json.

blakeembrey commented:
I'm happy to have the package replace Typings, I just haven't had the time, so I'll re-open the PR in case you'd like to tackle it.

devoto13 commented:
I've updated most of the packages. We need what to do with packages, which don't have available `@types` packages and don't bundle type definitions.

These are missing types:
- `tsconfig` (I see you just have migrated it)
- `v8flags`
- `proxyquire`
- `yn`

devoto13 commented on tsconfig.json:
```diff
@@ -1,7 +1,7 @@
   "compilerOptions": {
     "target": "es5",
-    "lib": ["es2015"],
+    "lib": ["es2015", "dom"],
```
`@types/react` depends on DOM APIs. We should either specify all types packages [explicitly](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types) in this tsconfig to prevent `react` from being included or add DOM APIs as it is done now.

blakeembrey replied:
I'd rather be specific than include DOM APIs since this project doesn't use React or the DOM directly.

devoto13 replied:
Okey, will do it than.

blakeembrey commented:
Sounds good. An alternative we can do is just to install from GitHub if we add `package.json` files to each of these projects.

devoto13 commented:
Good idea with installing from GitHub! Adding `package.json` was actually not necessary. 

Remaining type errors will require updating type definition for Node. Will work on it tomorrow.

blakeembrey commented on tsconfig.json:
```diff
@@ -13,10 +13,24 @@
     "sourceMap": true,
     "inlineSources": true,
-    "types": []
+    "types": [
```
You should only need to put global types here, not all of them, so it should work using just `node`.

devoto13 replied:
Also `mocha` is needed.

blakeembrey replied:
Right, for global test functions 👍  Good catch.

blakeembrey commented on package.json:
```diff
@@ -60,10 +64,18 @@
+    "@types/mkdirp": "^0.5.0",
+    "@types/node": "^6.0.88",
+    "@types/source-map-support": "^0.4.0",
+    "@types/yn": "types/npm-yn",
```
Can we put the commit hash for the GitHub installed dependencies? Otherwise we might get a major version that's incompatible later and we won't want that to break everything 😄

devoto13 replied:
Sure!

blakeembrey commented on src/types.d.ts:
```diff
@@ -0,0 +1 @@
+declare module 'tsconfig'
```
Released a major version of `tsconfig` now and we can update to it 👍

devoto13 commented:
I have prepared update for Node type definitions. Take a look [here](https://github.com/devoto13/DefinitelyTyped/commit/f2627839e9175a40665d65d161b6b8f5cca6e194). I think it worths bumping `@types/node` to 8.* to not mess up with type definitions for older versions and CI is running Node 8 as well. 

I decided to add extra Module class to avoid breaking people, who depend on `NodeModule` being an interface. And also unsure about including private variables/methods. Maybe we should access them by casting to `any` instead. What do you think?

devoto13 commented on src/index.ts:
```diff
@@ -12,7 +12,7 @@ import * as TS from 'typescript'
 
 const pkg = require('../package.json')
-const shouldDebug = yn(process.env.TS_NODE_DEBUG)
+const shouldDebug = yn(process.env.TS_NODE_DEBUG!)
```
Just to get it working for now. Type definitions fro `yn` should be updated instead as environment variables can be missing.

devoto13 commented on src/index.spec.ts:
```diff
@@ -79,6 +79,10 @@ describe('ts-node', function () {
 
     it('should throw errors', function (done) {
       exec(`${BIN_EXEC} -e "import * as m from './tests/module';console.log(m.example(123))"`, function (err) {
+        if (err === null) {
```
Since in newer Node type definitions `err` can be `null`. It was required to add check to satisfy `strictNullChecks`.

devoto13 commented:
Waiting for https://github.com/types/npm-yn/pull/17 to be merged and it should be ready for review. 

Let me know what you think about casting to `any` to access private methods. It's not recommended to add private APIs on DefinitelyTyped, so I added only public APIs.

blakeembrey commented:
> It's not recommended to add private APIs on DefinitelyTyped, so I added only public APIs.

That's a bummer, it is the node.js API and hasn't changed for years (at least since 0.8 for most of those, a couple since 4.0). How is one meant to type check between node versions for things like this?

I'm ok casting it to `any`, but would prefer pulling the required changes in locally.

devoto13 commented:
> That's a bummer, it is the node.js API and hasn't changed for years (at least since 0.8 for most of those, a couple since 4.0). How is one meant to type check between node versions for things like this?

I can see their reasons. Using private APIs is always "at your own risk". And even if these APIs don't change often, others do. So deciding which should be allowed and which should not gonna be a mess.

I have rebased and squashed commits. I can't pull in missing types locally because of https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19612. The problem is that it's not possible to augment symbol exported with `export =` syntax.

We can either merge "as is" and add missing types in a separate PR or wait for linked PR to be merged.
